util/genrest/goviewcreator.go

```diff
@@ -18,47 +18,130 @@ import (
+// current approach is to generate one file per service, with that file containing all the service's
+// RPCs. An additional file `genrest.go` is also created to register all these handlers with a
+// gorilla/mux dispatcher.
 func NewView(model *gomodel.Model) (*goview.View, error) {
-	view := goview.New(len(model.Service))
+	// TODO: Assert that all services live in the same proto package, which is currently the case and we currently assume.
+	namer := NewNamer()
 
-	for idxService, service := range model.Service {
-		file := view.Append(goview.NewFile("", strings.ToLower(service.ShortName)+".txt"))
-		file.P("AUTOGENERATED STUB shim %d: %s\n(This will be replaced by a generated Go source file)\n", idxService, service.FullName())
-		file.P("--- Imports ----------")
+	numServices := len(model.Service)
+	view := goview.New(numServices)
+	registered := []*registeredHandler{}
 
-		namer := NewNamer()
+	for idxService, service := range model.Service {
+		file := view.Append(goview.NewFile("", strings.ToLower(service.ShortName)+".go"))
+		file.P(license)
+		file.P("// DO NOT EDIT. This is an auto-generated file containing the REST handlers")
+		file.P("// for service #%d: %s.\n", idxService, service.FullName())
+		file.P("")
+		file.P("package genrest")
+		file.P("")
 
 		importStrings := make([]string, 0, len(service.Imports))
 		for _, spec := range service.Imports {
-			importStrings = append(importStrings, fmt.Sprintf("%s: %q", spec.Name, spec.Path))
+			importStrings = append(importStrings, fmt.Sprintf("%s %q", spec.Name, spec.Path))
 		}
 		sort.Strings(importStrings)
-		file.P(strings.Join(importStrings, "\n"))
 
-		file.P("\n--- Handlers ----------\n")
+		file.P("import (")
+		file.P(`  "context"`)
+		file.P(`  "net/http"`)
+		file.P("")
+		// TODO: Properly deal with imports once we actually use them in the code.
+		// file.P("  // %s", strings.Join(importStrings, "\n  // "))
+
+		// TODO: Get the following import from the data model, rather than hard-coding, once we deal with all imports.
+		file.P(`	  genprotopb "github.com/googleapis/gapic-showcase/server/genproto"`)
+
+		// TODO: Investigate why "google.golang.org/protobuf" doesn't work below
+		file.P(`  "github.com/golang/protobuf/jsonpb"`)
+		file.P(")")
+		file.P("")
 
 		for _, handler := range service.Handlers {
 			handlerName := namer.Get("Handle" + handler.GoMethod)
-			file.P("//                For :%s", handler.URIPattern)
-			file.P("// Match URIs of form :%s", matchingPath(handler.PathTemplate))
-			file.P("func %s(w http.ResponseWriter, r *http.Request) {", handlerName)
-			file.P("  // serialize from r to: var %s %s.%s", handler.RequestVariable, handler.RequestTypePackage, handler.RequestType)
-			file.P("  %s := %s(%s)", handler.ResponseVariable, handler.GoMethod, handler.RequestVariable)
-			file.P("  // serialize back to w\n")
+			pathMatch := matchingPath(handler.PathTemplate)
+			registered = append(registered, &registeredHandler{pathMatch, handlerName})
```

Comment:
(Not for this PR)
Should we have integrated file-diffing tests so we can verify the expected and actual file outputs? Please ignore if these are already present in the codebase/PR somewhere.

Reply:
Maybe. TBH, the main value of having golden files IMO is to make it easier for reviewers to see how code changes affect the output; that's why I'm including the relevant generated files in this PR. Maybe what we should do instead of golden file tests *per se* is add a presubmit check that requires that if these `server/genrest/` auto-generated files would be changed by the code, that they get submitted in the same PR (so that they're reviewed)